keras/backend/common/dtypes.py

```diff
@@ -0,0 +1,310 @@
+    elif backend.is_tensor(x):
+        dtype = backend.standardize_dtype(x.dtype)
+    # when inputing python dtype or tensor dtype
+    elif isinstance(x, type) and x in PYTHON_SCALAR_DTYPES:
+        dtype = PYTHON_SCALAR_DTYPES[x]
+    elif hasattr(x, "name"):
+        dtype = x.name
+    elif hasattr(x, "__str__") and "torch" in str(dtype):
+        dtype = str(x).split(".")[-1]
+    # when inputing dtype string
+    elif x in ALLOWED_DTYPES:
+        dtype = x
+    elif x in ["int", "float"]:
+        if x == "int":
+            dtype = "int64"
+        else:
+            dtype = "float64"
+
+    if dtype is None or dtype not in ALLOWED_DTYPES:
+        raise ValueError(f"Invalid dtype: {x}")
+    return (
+        canonicalize_dtype(backend.floatx(), dtype) if canonicalize else dtype
+    )
+
+
+_bool_types = ["bool"]
+_int_types = [
+    "uint8",
+    "uint16",
+    "uint32",
+    "uint64",
+    "int8",
+    "int16",
+    "int32",
+    "int64",
+]
+_float_types = ["bfloat16", "float16", "float32", "float64"]
+_complex_types = ["complex64", "complex128"]
+_weak_types = ["int", "float", "complex"]
+_weak_types_python = [int, float, complex]
+
+
+def _default_types(dtype):
+    if not isinstance(dtype, str):
+        raise TypeError(f"Invalid type: {type(dtype)}. Must be a str.")
+    if dtype == "bfloat16":  # special case for bfloat16
+        dtype_indicator = "f"
+    else:
+        dtype_indicator = dtype[:1]
+
+    # get backend.floatx precision (16, 32, 64)
+    precision = backend.floatx()[-2:]
+
+    if dtype_indicator == "b":
+        return "bool"
+    elif dtype_indicator == "i":
+        return "int" + precision
+    elif dtype_indicator == "u":
+        return "uint" + precision
+    elif dtype_indicator == "f":
+        return "float" + precision
+    # TODO: support complex types
+    else:
+        raise ValueError(f"Invalid dtype: {dtype} that is failed to be parsed")
+
+
+def _keras_core_type(dtype, weak_type):
+    """Return the keras core type for a dtype and weak type."""
+    if weak_type:
+        if dtype == "bool":
+            return dtype
+        if "float" in dtype:
+            return "float"
+        if "int" in dtype:
+            return "int"
+        # TODO: support complex types
+        # if "complex" in dtype:
+        #     return "complex"
+    return dtype
+
+
+def is_weakly_typed(x):
+    return type(x) in _weak_types_python
+
+
+def _dtype_and_weaktype(value):
+    """Return a (dtype, weak_type) tuple for the given input."""
+    return dtype(value), any(
+        value is typ for typ in _weak_types_python
+    ) or is_weakly_typed(value)
+
+
+def _type_promotion_lattice():
+    """
+    Return the type promotion lattice in the form of a DAG.
+    This DAG maps each type to its immediately higher type on the lattice.
+    """
+    (b1,) = _bool_types
+    (u1, u2, u4, u8, i1, i2, i4, i8) = _int_types
+    bf, f2, f4, f8 = _float_types
+    c4, c8 = _complex_types
```

Comment:
There are no plans to support complex numbers so you can drop all complex number references.

Reply:
Removed